src/app/clusters/scenes-server/scenes-server.cpp

```diff
@@ -0,0 +1,940 @@
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "scenes-server.h"
+
+#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/CommandHandlerInterface.h>
+#include <app/InteractionModelEngine.h>
+#include <app/clusters/scenes-server/SceneTableImpl.h>
+#include <app/server/Server.h>
+#include <app/util/attribute-storage.h>
+#include <credentials/GroupDataProvider.h>
+#include <lib/support/CommonIterator.h>
+#include <lib/support/Span.h>
+#include <lib/support/logging/CHIPLogging.h>
+#include <platform/PlatformManager.h>
+
+using SceneTableEntry   = chip::scenes::DefaultSceneTableImpl::SceneTableEntry;
+using SceneStorageId    = chip::scenes::DefaultSceneTableImpl::SceneStorageId;
+using SceneData         = chip::scenes::DefaultSceneTableImpl::SceneData;
+using HandlerContext    = chip::app::CommandHandlerInterface::HandlerContext;
+using ExtensionFieldSet = chip::scenes::ExtensionFieldSet;
+using GroupDataProvider = chip::Credentials::GroupDataProvider;
+using SceneTable        = chip::scenes::SceneTable<chip::scenes::ExtensionFieldSetsImpl>;
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace Scenes {
+
+ScenesServer ScenesServer::mInstance;
+
+ScenesServer & ScenesServer::Instance()
+{
+    return mInstance;
+}
+
+CHIP_ERROR ScenesServer::Init()
+{
+    // Prevents re-initializing
+    VerifyOrReturnError(!mIsInitialized, CHIP_ERROR_INCORRECT_STATE);
+
+    ReturnErrorOnFailure(chip::app::InteractionModelEngine::GetInstance()->RegisterCommandHandler(this));
+    VerifyOrReturnError(registerAttributeAccessOverride(this), CHIP_ERROR_INCORRECT_STATE);
+
+    mSceneTable = chip::scenes::GetSceneTableImpl();
+    ReturnErrorOnFailure(mSceneTable->Init(&chip::Server::GetInstance().GetPersistentStorage()));
+
+    mGroupProvider = Credentials::GetGroupDataProvider();
+
+    mFeatureFlags.Set(ScenesFeature::kSceneNames);
+    // The bit of 7 the NameSupport attribute indicates whether or not scene names are supported
+    //
+    // According to spec, bit 7 (Scene Names) MUST match feature bit 0 (Scene Names)
+    mNameSupport = 0x80;
+
+    // TODO: Make modifications to the CommandHandler and Transport layer to be able to forward the NodeId to clusters
+    mLastConfiguredBy.SetNull();
+    mIsInitialized = true;
+    return CHIP_NO_ERROR;
+}
+
+template <typename CommandData, typename ResponseType>
+void AddSceneParse(CommandHandlerInterface::HandlerContext & ctx, const CommandData & req, SceneTable * sceneTable,
+                   GroupDataProvider * groupProvider)
```

Comment:
The description of the class and the way it was used led me to believe this was the way to reduce reliance on old emberAf callbacks. Is this correct or was it a misunderstanding on my part?

Reply:
It's a way to reduce reliance on the emberAf callbacks for situations where people want to reduce the reliance.  But do we want to reduce it here?  It's not like we are trying to get rid of them in general; they were just a poor fit for some very specific and complicated use cases...

Reply:
Then it was a misunderstanding on my part about, I was under the impression we were slowly moving away from those where we could.

Reply:
Note also that if you use the static functions then unused things (because turned off in the config) will in fact be dropped by the compiler.  Whereas the CommandHandlerInterface approach as you noted in your other comments means all the code is always compiled in, including for commands that are not actually supported.

Reply:
Create issue [here](https://github.com/project-chip/connectedhomeip/issues/26901)